src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1616,6 +1616,88 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
 
 //------------------------------is_counted_loop--------------------------------
 bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  return do_is_counted_loop(x, loop, iv_bt) || is_counted_loop_with_speculative_long_limit(x, loop, iv_bt);
+}
+
+// When comparing an int incrementor with long limit, check if the limit is within int range. If so, treat it as an int
+// counted loop but deoptimize if the limit is out of int range.
+// This is common pattern with "for (int i =...; i < long_limit; ...)" where int "i" is implicitly promoted to long,
+// i.e., "(long) i < long_limit", and therefore making it not an int counted loop without this transformation.
+bool PhaseIdealLoop::is_counted_loop_with_speculative_long_limit(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  if (iv_bt != T_INT) {
+    return false;
+  }
+
+  Node* back_control = loop_exit_control(x, loop);
+  if (back_control == nullptr) {
+    return false;
+  }
+
+  Node* init_control = x->in(LoopNode::EntryControl);
+
+  // Make sure there is a parse predicate for us to insert the loop limit check.
```

Comment:
I usually write the predicate names capitalized to better highlight that these are well-defined names:
```suggestion
  // Make sure there is a Loop Limit Check Parse Predicate for us to insert the Loop Limit Check Predicate above it.
```